### ISSUE TITLE
ognl 3.4.3 string equals bug

### ISSUE THREAD

**Issue by hengyunabc** [orphan-oss/ognl#309]
https://github.com/hengyunabc/ognl34-bug-demo

## ognl 3.4.3 result

```
"aaa".equals(params[0].flowAttribute.getBxApp()), result: false
"aaa" == params[0].flowAttribute.getBxApp(), result: true
params[0].flowAttribute.getBxApp().equals("aaa"), result: true
```

## ognl 3.3.5 result

```
"aaa".equals(params[0].flowAttribute.getBxApp()), result: true
"aaa" == params[0].flowAttribute.getBxApp(), result: true
params[0].flowAttribute.getBxApp().equals("aaa"), result: true
```

**Comment by lukaszlenart** [comment]
Isn't this related to #74 and `USE_STRICTER_INVOCATION`? Please try to run your app using
 `-Dognl.UseStricterInvocation=false`

**Comment by lukaszlenart** [comment]
Either using `ognlContext.setRoot(advice);` or `Object object = Ognl.getValue(conditionExpress, advice);` solves the problem. Yet I'm not able to figure it out why your example fails.

**Comment by lukaszlenart** [comment]
Looks like it's due to #204

**Comment by lukaszlenart** [comment]
See the proposed fix, basically specifying `root` on `context` solves the problem, because due to changes in #204 `null` is returned instead.

**Comment by harawata** [comment]
Hi @lukaszlenart ,

I apologize if this misses the point, but the Javadoc of  [`Ognl#getValue(Object, OgnlContext, Object, Class<?>)`](https://github.com/orphan-oss/ognl/blob/58748354b1fd09cd66becd23e830fc2e85e55b04/src/main/java/ognl/Ognl.java#L386-L415) says...

> The default context is set for the given context and root via addDefaultContext().

But the call to `addDefaultContext()` is removed from `getValue()` and some other methods in [#193](https://github.com/orphan-oss/ognl/pull/193/files#diff-7c39755a1d9b668bf0eeaefe407b6c5b2a0a54ff3d73b6c669ce84502cd16b38L407).

I too am not fully understand, but could this cause the reported behavior change?

**Comment by lukaszlenart** [comment]
@harawata good catch, thanks! I will revert back to `addDefaultContext(root, context);` as I cannot recall what did I remove it
